tmt/hardware.py

```diff
@@ -0,0 +1,1088 @@
+        these constraints can also be compound constraints, we need to construct a cartesian
+        product of spans yielded by child constraints to include all possible combinations.
+
+        :param logger: logger to use for logging.
+        :param members: if specified, each span generated by this method is prepended with this
+            list.
+        :yields: all possible spans.
+        """
+
+        members = members or []
+
+        # List of non-compound constraints - we just slap these into every combination we generate
+        simple_constraints = [
+            constraint
+            for constraint in self.constraints
+            if not isinstance(constraint, CompoundConstraint)
+            ]
+
+        # Compound constraints - these we will ask to generate their spans, and we produce
+        # cartesian product from the output.
+        compound_constraints = [
+            constraint
+            for constraint in self.constraints
+            if isinstance(constraint, CompoundConstraint)
+            ]
+
+        for compounds in itertools.product(*[constraint.spans()
+                                           for constraint in compound_constraints]):
+            # Note that `product` returns an item for each iterable, and those items are lists,
+            # because that's what `spans()` returns. Use `sum` to linearize the list of lists.
+            yield members + sum(compounds, []) + simple_constraints
+
+
+@dataclasses.dataclass(repr=False)
+class Or(CompoundConstraint):
+    """
+    Represents constraints that are grouped in ``or`` fashion.
+    """
+
+    def __init__(self, constraints: Optional[List[ConstraintBase]] = None) -> None:
+        """
+        Hold constraints that are grouped in ``or`` fashion.
+
+        :param constraints: list of constraints to group.
+        """
+
+        super().__init__(any, constraints=constraints)
+
+    def spans(
+            self,
+            members: Optional[List[ConstraintBase]] = None
+            ) -> Iterator[List[ConstraintBase]]:
+        """
+        Generate all distinct spans covered by this constraint.
+
+        Since the ``any`` reducer allows any child constraints to be satisfied for the whole group
+        to evaluate as ``True``, it is trivial to generate spans - each child constraint shall
+        provide its own "branch", and there is no need for products or joins of any kind.
+
+        :param logger: logger to use for logging.
+        :param members: if specified, each span generated by this method is prepended with this
+            list.
+        :yields: all possible spans.
+        """
+
+        members = members or []
+
+        for constraint in self.constraints:
+            for span in constraint.spans():
+                yield members + span
+
+
+#
+# Constraint parsing
+#
+
+def ungroupify(fn: Callable[[SpecType], ConstraintBase]) -> Callable[[SpecType], ConstraintBase]:
```

Comment:
Please, add a docstring. Similarly below.

Reply:
Added in e7c0774a.